File: src/Events/messageCreate.ts
import {
  checkCooldown,
  formatError,
} from "../Core/commandUtils.js";
import {
  sendPrefixCommandUsage,
  sendPrefixCommandError,
} from "../Core/prefixCommandWebhook.js";
import { sendErrorToWebhook } from "../Core/errorWebhook.js";
import type { ExtendedClient, EventFile } from "../types.js";

export default {
  name: "messageCreate",
  async execute(message: any, client: ExtendedClient) {
    if (message.author.bot || !message.guild) return;
    if (!message.content.startsWith(client.config.prefix)) return;

    const args = message.content
      .slice(client.config.prefix.length)
      .trim()
      .split(/ +/);
    const commandName = args.shift().toLowerCase();
    const command = client.prefixCommands.get(commandName);

    if (!command) return;

    const cooldownResult = checkCooldown(
      message.author.id,
      commandName,
      command.cooldown || 3000
    );
    if (cooldownResult.onCooldown) {
      return message.reply(
        `⏰ Please wait ${cooldownResult.timeLeft} seconds before using this command again.`
      );
    }

    sendPrefixCommandUsage(message.author, commandName, message.guild.name);

    try {
      await command.run(client, message, args);
    } catch (error) {
      const err = error as Error;

      sendPrefixCommandError(err, {
        commandName: commandName,
        user: message.author,
        guild: message.guild,
      });

      sendErrorToWebhook(err, {
        commandName: commandName,
        user: message.author,
        guild: message.guild,
      });

      if (!message.replied) {
        try {
          await message.reply(
            "❌ Error executing command! Please try again later."
          );
        } catch {
          // silent
        }
      }
    }
  },
} satisfies EventFile;

File: go/events/message_create.go
package events

import (
	"fmt"
	"strings"

	"github.com/RealMtrx/Discord-Handler-Go/bot"
	"github.com/RealMtrx/Discord-Handler-Go/core"
	"github.com/bwmarrin/discordgo"
)

func MessageCreate(s *discordgo.Session, m *discordgo.MessageCreate) {
	if m.Author.Bot || m.GuildID == "" {
		return
	}

	prefix := bot.App.Config.Prefix
	if !strings.HasPrefix(m.Content, prefix) {
		return
	}

	args := strings.Fields(strings.TrimPrefix(m.Content, prefix))
	if len(args) == 0 {
		return
	}

	cmdName := strings.ToLower(args[0])
	cmd, exists := bot.App.PrefixCommands[cmdName]
	if !exists {
		return
	}

	onCooldown, remaining := core.Cooldowns.Check(m.Author.ID, cmdName, 3000)
	if onCooldown {
		s.ChannelMessageSend(m.ChannelID,
			"⏰ Please wait "+fmt.Sprintf("%d", remaining)+" seconds before using this command again.")
		return
	}

	go cmd.Handler(s, m, args[1:])
}
